File: diagnostic_aggregator/src/diagnostic_aggregator/aggregator.py
#!/usr/bin/env python
# Software License Agreement (BSD License)
#
# Copyright (c) 2008, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of the Willow Garage nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#

# Author: Kevin Watts

## Aggregates diagnostics from robot and republishes to /diagnostics_agg

PKG = 'diagnostic_aggregator'
import roslib; roslib.load_manifest(PKG)

import rospy

from xml.dom import minidom

from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus, KeyValue

import traceback

import sys

import threading
from threading import Timer

from diagnostic_aggregator.generic_analyzer import GenericAnalyzer

stat_dict = { 0: 'OK', 1: 'Warning', 2: 'Error' }

def key_comp(a, b):
    return cmp(a.key, b.key)

##\brief Creates analyzer class from XML snippet
##
## Dyanamically loads analyzer class from given package, file and 
## type in XML snippet. Must be able to load manifest of give package
## and 
##\param xml_node mindom node : XML node to choose, initialize analyzer
##\param global_prefix str : Prefix of all analyzers for this aggregator
def create_analyzer(xml_node, global_prefix):
    type   = xml_node.attributes['type'].value
    pkg    = xml_node.attributes['pkg'].value
    file   = xml_node.attributes['file'].value
    local_prefix = xml_node.attributes['prefix'].value
    
    prefix = '/'.join([global_prefix, local_prefix])

    if pkg != PKG:
        roslib.load_manifest(pkg)
        
    import_str = '%s.%s' % (pkg, file)
    __import__(import_str)

    try:
        pypkg = sys.modules[import_str]
        my_analyzer = getattr(pypkg, type) # Type of analyzer
    except KeyError:
        rospy.logfatal("Cannot load package %s, module %s" % (pkg, file))
        sys.exit(-1)

    try:
        # Create dynamically
        analyzer = my_analyzer(xml_node, prefix)
    except:
        rospy.logfatal('Failed to create analyzer. %s' % traceback.format_exc())
        sys.exit(-1)

    # Check that it has attributes analyze, topic
        
    return analyzer, prefix

##\brief Aggregates /diagnostics topic and republishes on /diagnostics_agg
## 
## DiagnosticAggregator ia an aggregator that creates "analyzers" to 
## analyze, store and track DiagnosticStatus messages. It publishes the 
## aggregated data on /diagnostics_agg as a DiagnosticArray. It prepends the 
## given prefix to all aggregated DiagnosticStatus message names.
## 
## The XML snippet to initialize the aggregator contains the necessary 
## data to dynamically load the analyzers.
##\verbatim
## <aggregator>
##  <analyzer type="AnalyzerClassName" pkg="analyzer_package"
##            file="analyzer_file" prefix="Prefix" >
##   <!-- Anything else to initialize the analyzer -->
##  </analyzer>
## <!-- Other analyzers -->
## </aggregator>
##\endverbatim
## The analyzer will prepend the prefix "/aggregator_prefix/Prefix/" to 
## all topics it analyzes. 
class DiagnosticAggregator:
    ##\param prefix str : Prepended to all processed DiagnosticStatus names
    ##\param xml_doc minidom node : XML node for analyzer creation
    def __init__(self, prefix, xml_doc):
        self._analyzers = {}
        self._msgs = []

        self._mutex = threading.Lock()

        if not prefix.startswith('/'):
            prefix = '/' + prefix
        self._prefix = prefix
        
        analyzers = xml_doc.getElementsByTagName('analyzer')
        for analyzer in analyzers:
            new_analyzer, second_parent = create_analyzer(analyzer, self._prefix)
            self._analyzers[new_analyzer] = second_parent

        self._remainder_analyzer = GenericAnalyzer(None, 
                                                   self._prefix + '/Other', True)

        rospy.Subscriber('/diagnostics', DiagnosticArray, self.diag_callback)

        self._agg_pub = rospy.Publisher('/diagnostics_agg', DiagnosticArray)

    ##\brief Callback for /diagnostics subscription
    def diag_callback(self, array):
        self._mutex.acquire()
        
        for msg in array.status:
            self._msgs.append(msg)
        self._mutex.release()

    ##\brief Calls analyzers, publishes /diagnostics_agg 
    ## 
    ## Publishes on /diagnostics_agg, fills out "Header" or first parent
    ## node with data on the child nodes below it. Calls analyzers, and 
    ## calls remainder analyzer on the remainder. 
    def publish_data(self):
        self._mutex.acquire()

        # Global status
        header_status = DiagnosticStatus()
        header_status.name = self._prefix
        header_status.level = 0
        header_status.message = 'OK'
        header_status.values = []

        array = DiagnosticArray()
        
        array.status = [ header_status ]

        msg_dict = {}
        for msg in self._msgs:
            msg_dict[msg] = 0

        for analyzer, second_parent in self._analyzers.iteritems():
            data  = analyzer.analyze(msg_dict)
            
            # Add branch names, status to root item
            for item in data:
                if item.name == second_parent:
                    nice_name = str(second_parent.split('/')[-1])
                    header_status.values.append(KeyValue(nice_name, item.message))
                    break
            array.status.extend(data)
            
        # Process remaining data
        other_data = self._remainder_analyzer.analyze(msg_dict)
        for item in other_data:
            if item.name == self._prefix + '/Other':
                header_status.values.append(KeyValue('Other', item.message))
                break
         
        array.status.extend(other_data)

        # Sort header status by keys
        header_status.values.sort(key_comp)

        for sub_status in array.status:
            header_status.level = max(sub_status.level, header_status.level)
        if header_status.level == 3:
            header_status.level = 2 # Report stale as error
        header_status.message = stat_dict[header_status.level]
        
        self._agg_pub.publish(array)
        self._msgs = []

        self._mutex.release()

File: diagnostic_aggregator/src/aggregator.cpp
/*********************************************************************
 * Software License Agreement (BSD License)
 *
 *  Copyright (c) 2009, Willow Garage, Inc.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of the Willow Garage nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/

/**! \author Kevin Watts */
/**! \author Arne Nordmann */

#include "diagnostic_aggregator/aggregator.hpp"

#include <chrono>
#include <map>
#include <memory>
#include <mutex>
#include <string>
#include <vector>

namespace diagnostic_aggregator
{
using std::placeholders::_1;
using std::placeholders::_2;
using std::placeholders::_3;

using diagnostic_msgs::msg::DiagnosticArray;
using diagnostic_msgs::msg::DiagnosticStatus;

/**
 * @todo(anordman): make aggregator a lifecycle node.
 */
Aggregator::Aggregator()
: Aggregator(rclcpp::NodeOptions()) {}

Aggregator::Aggregator(rclcpp::NodeOptions options)
: n_(std::make_shared<rclcpp::Node>(
      "analyzers", "",
      options.allow_undeclared_parameters(true).
      automatically_declare_parameters_from_overrides(true))),
  logger_(rclcpp::get_logger("Aggregator")),
  pub_rate_(1.0),
  history_depth_(1000),
  clock_(n_->get_clock()),
  base_path_(""),
  critical_(false),
  publish_values_(true),
  last_top_level_state_(DiagnosticStatus::STALE)
{
  RCLCPP_DEBUG(logger_, "constructor");
  initAnalyzers();

  diag_sub_ = n_->create_subscription<DiagnosticArray>(
    "/diagnostics", rclcpp::SystemDefaultsQoS().keep_last(history_depth_),
    std::bind(&Aggregator::diagCallback, this, _1));
  agg_pub_ = n_->create_publisher<DiagnosticArray>("/diagnostics_agg", 1);
  toplevel_state_pub_ =
    n_->create_publisher<DiagnosticStatus>("/diagnostics_toplevel_state", 1);

  int publish_rate_ms = 1000 / pub_rate_;
  publish_timer_ = n_->create_wall_timer(
    std::chrono::milliseconds(publish_rate_ms),
    std::bind(&Aggregator::publishData, this));

  param_cb_handle_ = n_->add_on_set_parameters_callback(
    std::bind(&Aggregator::onParametersSet, this, _1));
}

rcl_interfaces::msg::SetParametersResult Aggregator::onParametersSet(
  const std::vector<rclcpp::Parameter> & parameters)
{
  // Check if any of the incoming parameters are new. If so, flag for reinitialization.
  // The method publishData() will pick it up on the next publish cycle and call initAnalyzers().
  for (const auto & p : parameters) {
    if (!n_->has_parameter(p.get_name())) {
      reinit_needed_.store(true);
      break;
    }
  }

  rcl_interfaces::msg::SetParametersResult result;
  result.successful = true;
  return result;
}

void Aggregator::initAnalyzers()
{
  bool other_as_errors = false;

  std::map<std::string, rclcpp::Parameter> parameters;
  if (!n_->get_parameters("", parameters)) {
    RCLCPP_ERROR(logger_, "Couldn't retrieve parameters.");
  }
  RCLCPP_DEBUG(logger_, "Retrieved %zu parameter(s).", parameters.size());

  for (const auto & param : parameters) {
    if (param.first.compare("pub_rate") == 0) {
      pub_rate_ = param.second.as_double();
    } else if (param.first.compare("path") == 0) {
      // Leading slash when path is not empty
      if (!param.second.as_string().empty()) {
        base_path_.append("/");
      }
      base_path_.append(param.second.as_string());
    } else if (param.first.compare("other_as_errors") == 0) {
      other_as_errors = param.second.as_bool();
    } else if (param.first.compare("history_depth") == 0) {
      history_depth_ = param.second.as_int();
    } else if (param.first.compare("critical") == 0) {
      critical_ = param.second.as_bool();
    } else if (param.first.compare("publish_values") == 0) {
      publish_values_ = param.second.as_bool();
    }
  }
  RCLCPP_DEBUG(logger_, "Aggregator publication rate configured to: %f", pub_rate_);
  RCLCPP_DEBUG(logger_, "Aggregator base path configured to: %s", base_path_.c_str());
  RCLCPP_DEBUG(
    logger_, "Aggregator other_as_errors configured to: %s", (other_as_errors ? "true" : "false"));
  RCLCPP_DEBUG(
    logger_, "Aggregator critical publisher configured to: %s", (critical_ ? "true" : "false"));
  RCLCPP_DEBUG(
    logger_, "Aggregator publish_values configured to: %s", (publish_values_ ? "true" : "false"));

  {  // lock the mutex while analyzer_group_ and other_analyzer_ are being updated
    std::lock_guard<std::mutex> lock(mutex_);
    analyzer_group_ = std::make_unique<AnalyzerGroup>();
    if (!analyzer_group_->init(base_path_, "", n_)) {
      RCLCPP_ERROR(logger_, "Analyzer group for diagnostic aggregator failed to initialize!");
    }

    // Last analyzer handles remaining data
    other_analyzer_ = std::make_unique<OtherAnalyzer>(other_as_errors);
    other_analyzer_->init(base_path_);  // This always returns true
  }
}

void Aggregator::checkTimestamp(const DiagnosticArray::SharedPtr diag_msg)
{
  RCLCPP_DEBUG(logger_, "checkTimestamp()");
  if (diag_msg->header.stamp.sec != 0) {
    return;
  }

  std::string stamp_warn = "No timestamp set for diagnostic message. Message names: ";
  std::vector<DiagnosticStatus>::const_iterator it;
  for (it = diag_msg->status.begin(); it != diag_msg->status.end(); ++it) {
    if (it != diag_msg->status.begin()) {
      stamp_warn += ", ";
    }
    stamp_warn += it->name;
  }

  auto result = ros_warnings_.insert(stamp_warn);
  if (result.second) {
    RCLCPP_WARN(logger_, "%s", stamp_warn.c_str());
  }
}

void Aggregator::diagCallback(const DiagnosticArray::SharedPtr diag_msg)
{
  RCLCPP_DEBUG(logger_, "diagCallback()");
  checkTimestamp(diag_msg);

  bool analyzed = false;
  bool immediate_report = false;
  {  // lock the whole loop to ensure nothing in the analyzer group changes during it.
    std::lock_guard<std::mutex> lock(mutex_);
    for (auto j = 0u; j < diag_msg->status.size(); ++j) {
      analyzed = false;
      auto item = std::make_shared<StatusItem>(&diag_msg->status[j]);

      if (analyzer_group_->match(item->getName())) {
        analyzed = analyzer_group_->analyze(item);
      }

      if (!analyzed) {
        other_analyzer_->analyze(item);
      }

      // In case there is a degraded state, publish immediately
      if (critical_ && item->getLevel() > last_top_level_state_) {
        immediate_report = true;
      }
    }
  }

  if (immediate_report) {
    publishData();
  }
}

Aggregator::~Aggregator()
{
  RCLCPP_DEBUG(logger_, "destructor");
}

void Aggregator::publishData()
{
  RCLCPP_DEBUG(logger_, "publishData()");

  // Check if reinitialization is needed because new parameters have been set.
  if (reinit_needed_.exchange(false)) {
    base_path_ = "";
    initAnalyzers();
  }

  DiagnosticArray diag_array;
  DiagnosticStatus diag_toplevel_state;
  diag_toplevel_state.name = "toplevel_state";
  diag_toplevel_state.level = DiagnosticStatus::STALE;
  int max_level = -1;
  uint8_t max_level_without_stale = 0;
  int non_ok_status_depth = 0;
  std::shared_ptr<DiagnosticStatus> msg_to_report;

  std::vector<std::shared_ptr<DiagnosticStatus>> processed;
  {
    std::lock_guard<std::mutex> lock(mutex_);
    processed = analyzer_group_->report();
  }
  for (const auto & msg : processed) {
    diag_array.status.push_back(*msg);
    const auto depth = std::count(msg->name.begin(), msg->name.end(), '/');

    if (msg->level > max_level) {
      max_level = msg->level;
      non_ok_status_depth = depth;
      msg_to_report = msg;
    }
    if (msg->level == max_level && depth > non_ok_status_depth) {
      // On non okay diagnostics also copy the deepest message to toplevel state
      non_ok_status_depth = depth;
      msg_to_report = msg;
    }
    if (
      msg->level > max_level_without_stale &&
      msg->level != diagnostic_msgs::msg::DiagnosticStatus::STALE)
    {
      max_level_without_stale = msg->level;
    }
  }

  std::vector<std::shared_ptr<DiagnosticStatus>> processed_other =
    other_analyzer_->report();
  for (const auto & msg : processed_other) {
    diag_array.status.push_back(*msg);
    const auto depth = std::count(msg->name.begin(), msg->name.end(), '/');

    if (msg->level > max_level) {
      max_level = msg->level;
      non_ok_status_depth = depth;
      msg_to_report = msg;
    }
    if (msg->level == max_level && depth > non_ok_status_depth) {
      // On non okay diagnostics also copy the deepest message to toplevel state
      non_ok_status_depth = depth;
      msg_to_report = msg;
    }
    if (
      msg->level > max_level_without_stale &&
      msg->level != diagnostic_msgs::msg::DiagnosticStatus::STALE)
    {
      max_level_without_stale = msg->level;
    }
  }

  // When a non-ok item was found, surface the offender via message/hardware_id/values
  // but keep name stable as "toplevel_state" to avoid breaking downstream consumers
  if (max_level > DiagnosticStatus::OK && msg_to_report) {
    diag_toplevel_state.message = msg_to_report->name + ": " + msg_to_report->message;
    diag_toplevel_state.hardware_id = msg_to_report->hardware_id;
    diag_toplevel_state.values = msg_to_report->values;
  }

  // If "publish_values" is false, clear all values
  if (!publish_values_) {
    for (auto & status : diag_array.status) {
      status.values.clear();
    }
  }

  diag_array.header.stamp = clock_->now();
  agg_pub_->publish(diag_array);

  if (max_level_without_stale > DiagnosticStatus::OK) {
    diag_toplevel_state.level = max_level_without_stale;
  } else if (max_level == diagnostic_msgs::msg::DiagnosticStatus::STALE) {
    diag_toplevel_state.level = diagnostic_msgs::msg::DiagnosticStatus::STALE;
  } else if (max_level < 0) {
    diag_toplevel_state.level = diagnostic_msgs::msg::DiagnosticStatus::STALE;
  } else {
    diag_toplevel_state.level = DiagnosticStatus::OK;
  }

  last_top_level_state_ = diag_toplevel_state.level;

  toplevel_state_pub_->publish(diag_toplevel_state);
}

rclcpp::Node::SharedPtr Aggregator::get_node() const
{
  RCLCPP_DEBUG(logger_, "get_node()");
  return this->n_;
}

}  // namespace diagnostic_aggregator
